topostats/tracing/disordered_tracing.py

```diff
@@ -0,0 +1,793 @@
+
+    @staticmethod
+    def remove_touching_edge(skeleton: npt.NDArray) -> npt.NDArray:
+        """
+        Remove any skeleton points touching the border (to prevent errors later).
+
+        Parameters
+        ----------
+        skeleton : npt.NDArray
+            A binary array where touching clusters of 1's become 0's if touching the edge of the array.
+
+        Returns
+        -------
+        npt.NDArray
+            Skeleton without points touching the border.
+        """
+        for edge in [skeleton[0, :-1], skeleton[:-1, -1], skeleton[-1, 1:], skeleton[1:, 0]]:
+            uniques = np.unique(edge)
+            for i in uniques:
+                skeleton[skeleton == i] = 0
+        return skeleton
+
+    def smooth_mask(
+        self,
+        grain: npt.NDArray,
+        dilation_iterations: int = 2,
+        gaussian_sigma: float | int = 2,
+        holearea_min_max: tuple[int | float | None] = (0, None),
+    ) -> npt.NDArray:
+        """
+        Smooth a grain mask based on the lower number of binary pixels added from dilation or gaussian.
+
+        This method ensures gaussian smoothing isn't too aggressive and covers / creates gaps in the mask.
+
+        Parameters
+        ----------
+        grain : npt.NDArray
+            Numpy array of the grain mask.
+        dilation_iterations : int
+            Number of times to dilate the grain to smooth it. Default is 2.
+        gaussian_sigma : float | None
+            Gaussian sigma value to smooth the grains after an Otsu threshold. If None, defaults to 2.
+        holearea_min_max : tuple[float | int | None]
+            Tuple of minimum and maximum hole area (in nanometers) to replace from the original mask into the smoothed
+            mask.
+
+        Returns
+        -------
+        npt.NDArray
+            Numpy array of smmoothed image.
+        """
+        dilation = ndimage.binary_dilation(grain, iterations=dilation_iterations).astype(np.int32)
+        gauss = filters.gaussian(grain, sigma=gaussian_sigma)
+        gauss = np.where(gauss > filters.threshold_otsu(gauss) * 1.3, 1, 0)
+        gauss = gauss.astype(np.int32)
+        # Add hole to the smooth mask conditional on smallest pixel difference for dilation or the Gaussian smoothing.
+        if dilation.sum() > gauss.sum():
+            LOGGER.debug(f"[{self.filename}] : smoothing done by gaussian {gaussian_sigma}")
+            return self.re_add_holes(grain, gauss, holearea_min_max)
+        LOGGER.debug(f"[{self.filename}] : smoothing done by dilation {dilation_iterations}")
+        return self.re_add_holes(grain, dilation, holearea_min_max)
+
+
+def trace_image_disordered(  # pylint: disable=too-many-arguments,too-many-locals
+    image: npt.NDArray,
+    grains_mask: npt.NDArray,
+    filename: str,
+    pixel_to_nm_scaling: float,
+    min_skeleton_size: int,
+    mask_smoothing_params: dict,
+    skeletonisation_params: dict,
+    pruning_params: dict,
+    pad_width: int = 1,
+) -> dict:
```

Comment:
Typehint here is incorrect it should, I think, be

```suggestion
) -> tuple[dict, df, dict, df]:
```